lighthouse-treemap/app/src/util.js

```diff
@@ -64,8 +117,153 @@ class TreemapUtil {
+  }
+
+  /**
+   * Given a list of items, return a function (a hasher) that will map keys to an item.
+   * When a key is seen for the first time, the item returned is cached and will always
+   * be returned for the same key.
+   * The hash function is stable and deterministic, so the same key->item mapping will be
+   * produced given the same call order.
+   * @template T
+   * @param {T[]} items
+   * @return {(key: string) => T|undefined}
+   */
+  static stableHasher(items) {
+    // Clone.
+    items = [...items];
+
+    /** @type {Map<string, T>} */
+    const assignedItems = new Map();
+    return key => {
+      // Key has already been assigned an item.
+      if (assignedItems.has(key)) return assignedItems.get(key);
+
+      // Ran out of items.
+      if (items.length === 0) return;
+
+      // Select a random item using a stable hash.
+      const hash = [...key].reduce((acc, char) => acc + char.charCodeAt(0), 0);
+      const [assignedItem] = items.splice(hash % items.length, 1);
+      assignedItems.set(key, assignedItem);
+
+      return assignedItem;
+    };
+  }
+
+  /**
+   * @param {number} h
+   * @param {number} s
+   * @param {number} l
+   */
+  static hsl(h, s, l) {
+    return `hsl(${h}, ${s}%, ${l}%)`;
+  }
+
+  /**
+   * Brilliant code by akinuri
+   * https://stackoverflow.com/a/39147465
+   * @param {number} r
+   * @param {number} g
+   * @param {number} b
+   */
+  static rgb2hue(r, g, b) {
+    r /= 255;
+    g /= 255;
+    b /= 255;
+
+    const max = Math.max(r, g, b);
+    const min = Math.min(r, g, b);
+    const c = max - min;
+    let hue = 0;
+    let segment;
+    let shift;
+    if (c === 0) {
+      hue = 0;
+    } else {
+      switch (max) {
+        case r:
+          segment = (g - b) / c;
+          shift = 0 / 60; // R° / (360° / hex sides)
+          if (segment < 0) { // hue > 180, full rotation
+            shift = 360 / 60; // R° / (360° / hex sides)
+          }
+          hue = segment + shift;
+          break;
+        case g:
+          segment = (b - r) / c;
+          shift = 120 / 60; // G° / (360° / hex sides)
+          hue = segment + shift;
+          break;
+        case b:
+          segment = (r - g) / c;
+          shift = 240 / 60; // B° / (360° / hex sides)
+          hue = segment + shift;
+          break;
+      }
+    }
+    return hue * 60; // hue is in [0,6], scale it up
+  }
 }
 
+// From DevTools:
+// https://cs.chromium.org/chromium/src/third_party/devtools-frontend/src/front_end/quick_open/CommandMenu.js?l=255&rcl=ad5c586c30a6bc55962b7a96b0533911c86bd4fc
+TreemapUtil.COLOR_HUES = [
+  '#F44336', '#E91E63', '#9C27B0', '#673AB7', '#3F51B5', '#03A9F4',
+  '#00BCD4', '#009688', '#4CAF50', '#8BC34A', '#CDDC39', '#FFC107',
+  '#FF9800', '#FF5722', '#795548', '#9E9E9E', '#607D8B',
+].map(hex => {
+  const hexParts = hex.slice(1).split(/(..)/).filter(Boolean);
+  const [r, g, b] = hexParts.map(part => parseInt(part, 16));
+  return TreemapUtil.rgb2hue(r, g, b);
```

Comment:
** enter some diatribe from paul on how HSL > HEX again ie. https://mothereffinghsl.com/ **  (but yes i understand the devtools source of truth palette is expressed in hex. because pavel.)

but anyway.. since all you want is the hue value... let's not do that calculation on the client?? 

might as well hardcode these:


```js
TreemapUtil.COLOR_HUES

[
  4.105263157894738,
  339.60591133004925,
  291.24087591240874,
  261.6,
  230.84745762711864,
  198.67219917012446,
  186.79245283018867,
  174.4,
  122.42424242424241,
  87.76859504132231,
  65.52147239263803,
  45,
  35.76470588235294,
  14.389140271493213,
  15.918367346938773,
  0,
  199.53488372093022
]
```